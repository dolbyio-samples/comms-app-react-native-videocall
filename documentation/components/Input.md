# Input

The Input component is responsible for interactive controls for forms in order to accept data from the user.

## Props

| Name                   | Type                                 | Default | Description                                                                 |
| ---------------------- | ------------------------------------ | ------- | --------------------------------------------------------------------------- |
| `value`                | string                               | -       | The value of the input element that is required for a controlled component. |
| `label` ?              | string                               | -       | The text displayed above the input element.                                 |
| `borderColor` ?        | ColorKey                             | -       | The color of the border.                                                    |
| `invalidBorderColor` ? | ColorKey                             | -       | The color of the input element that contains an invalid value.              |
| `labelColor` ?         | ColorKey                             | -       | The color of text that is displayed in the label.                           |
| `labelBackground` ?    | ColorKey                             | -       | The background color of the label.                                          |
| `textColor` ?          | ColorKey                             | -       | The color of the input element.                                             |
| `validation` ?         | { valid: boolean; message?: string } | -       | The validation of the provided value.                                       |
| `onChangeText`         | function                             | -       | A callback which pass back the updated `text` value as an argument          |
| `testID` ?             | string                               | -       | The unique E2E test handler.                                                |

## Examples

### React

```javascript
return <Input label="Your Name" value={name} onChange={(event) => setName(event.target.value)} />;
```
