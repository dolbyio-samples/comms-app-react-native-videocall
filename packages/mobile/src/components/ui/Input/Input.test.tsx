import { defaultTheme as theme } from '@dolbyio/comms-uikit-common';
// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent } from '@testing-library/react-native';
import React from 'react';
import { View, TextInput } from 'react-native';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render, mergeStyles } from '../../../utils/test-utils';

import Input from './Input';

const { colors } = theme;

const testID = 'testID';
const label = 'testLabel';
const labelColor = 'red.400';
const labelBackground = 'grey.800';
const textColor = 'purple.400';

const onChange = () => {
  console.log('test');
};

describe('Input component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<Input testID={testID} value="" />);
    expect(getByTestId(testID)).not.toBeNull();
  });

  test('can change border color when text input is in focus', () => {
    const { getByTestId, rerender } = render(<Input testID={testID} value="foo" />);
    const element = getByTestId(testID);
    const viewElmList = element.findAllByType(View);
    const txtInputElm = element.findByType(TextInput);
    // fire focus event on text input element
    fireEvent(txtInputElm, 'focus');
    rerender(<Input testID={testID} value="foo" />);
    let styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.primary[400]);

    // fire blur event on text input element
    fireEvent(txtInputElm, 'blur');
    rerender(<Input testID={testID} value="foo" />);
    styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.grey[100]);
  });

  test('can use infoError as border color, when valid is false', () => {
    const validation = {
      valid: false,
    };
    const { getByTestId, rerender } = render(<Input testID={testID} validation={validation} value="foo" />);
    const element = getByTestId(testID);
    const viewElmList = element.findAllByType(View);
    const txtInputElm = element.findByType(TextInput);

    let styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.infoError);

    // fire focus event on text input element
    fireEvent(txtInputElm, 'focus');
    rerender(<Input testID={testID} validation={validation} value="foo" />);
    styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.infoError);
  });

  test('can change border color when text input is in focus, when valid is true', () => {
    const validation = {
      valid: true,
    };
    const { getByTestId, rerender } = render(<Input testID={testID} validation={validation} value="foo" />);
    const element = getByTestId(testID);
    const viewElmList = element.findAllByType(View);
    const txtInputElm = element.findByType(TextInput);
    // fire focus event on text input element
    fireEvent(txtInputElm, 'focus');
    rerender(<Input testID={testID} value="foo" />);
    let styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.primary[400]);

    // fire blur event on text input element
    fireEvent(txtInputElm, 'blur');
    rerender(<Input testID={testID} validation={validation} value="foo" />);
    styles = mergeStyles(viewElmList[1].props.style);
    expect(styles.borderColor).toBe(colors.grey[100]);
  });

  test('can handle input text and invoke onChange method', () => {
    const mockOnClick = jest.fn();
    const { getByTestId } = render(<Input testID={testID} value="foo" onChange={mockOnClick} />);
    const element = getByTestId(testID);
    const txtInputElm = element.findByType(TextInput);
    fireEvent(txtInputElm, 'onChange', '123');
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });
});

describe('Input component snapshot', () => {
  test('should match default param configuration', () => {
    const tree = renderer.create(<Input testID={testID} value="foo" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match with validation.message is set & validation.valid is true & value.length > 0', () => {
    const validation = {
      valid: true,
      message: 'validationMsg',
    };
    const tree = renderer
      .create(
        <Input
          label={label}
          labelColor={labelColor}
          textColor={textColor}
          labelBackground={labelBackground}
          validation={validation}
          testID={testID}
          value="foo"
          onChange={onChange}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match with validation.message is set & validation.valid is false & value.length == 0', () => {
    const validation = {
      valid: false,
      message: 'validationMsg',
    };
    const tree = renderer
      .create(
        <Input
          label={label}
          labelColor={labelColor}
          textColor={textColor}
          labelBackground={labelBackground}
          validation={validation}
          testID={testID}
          value=""
          onChange={onChange}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should match with validation.message is not set', () => {
    const validation = {
      valid: true,
    };
    const tree = renderer
      .create(
        <Input
          label={label}
          labelColor={labelColor}
          textColor={textColor}
          labelBackground={labelBackground}
          validation={validation}
          testID={testID}
          value=""
          onChange={onChange}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
