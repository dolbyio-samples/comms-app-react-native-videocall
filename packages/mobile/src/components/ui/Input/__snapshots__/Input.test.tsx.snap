// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component snapshot should match default param configuration 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
      "marginTop": 16,
    }
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "elevation": 1,
          "marginStart": 16,
          "position": "absolute",
          "top": -8,
          "zIndex": 1,
        },
        Object {
          "backgroundColor": undefined,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#6A6A6D",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 4,
          "borderWidth": 2,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingTop": 12,
          "zIndex": 0,
        },
        Object {
          "borderColor": "#E0E0E5",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "color": "black",
          "fontFamily": "Roboto",
        }
      }
      value="foo"
    />
  </View>
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "elevation": 1,
        "position": "absolute",
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 20,
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "overflow": "hidden",
          "width": 40,
        }
      }
    >
      <svg
        data-file-name="SvgClose"
        fill="#14141A"
        height={10}
        testID="Icon"
        width={10}
      />
    </View>
  </View>
</View>
`;

exports[`Input component snapshot should match with validation.message is not set 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
      "marginTop": 16,
    }
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "elevation": 1,
          "marginStart": 16,
          "position": "absolute",
          "top": -8,
          "zIndex": 1,
        },
        Object {
          "backgroundColor": "grey.800",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    >
      testLabel
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 4,
          "borderWidth": 2,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingTop": 12,
          "zIndex": 0,
        },
        Object {
          "borderColor": "#E0E0E5",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "color": "#AA33FF",
          "fontFamily": "Roboto",
        }
      }
      value=""
    />
  </View>
</View>
`;

exports[`Input component snapshot should match with validation.message is set & validation.valid is false & value.length == 0 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
      "marginTop": 16,
    }
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "elevation": 1,
          "marginStart": 16,
          "position": "absolute",
          "top": -8,
          "zIndex": 1,
        },
        Object {
          "backgroundColor": "grey.800",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    >
      testLabel
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 4,
          "borderWidth": 2,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingTop": 12,
          "zIndex": 0,
        },
        Object {
          "borderColor": "#E52222",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "color": "#AA33FF",
          "fontFamily": "Roboto",
        }
      }
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 8,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#E52222",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    >
      validationMsg
    </Text>
  </View>
</View>
`;

exports[`Input component snapshot should match with validation.message is set & validation.valid is true & value.length > 0 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
      "marginTop": 16,
    }
  }
  testID="testID"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "elevation": 1,
          "marginStart": 16,
          "position": "absolute",
          "top": -8,
          "zIndex": 1,
        },
        Object {
          "backgroundColor": "grey.800",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    >
      testLabel
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 4,
          "borderWidth": 2,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingTop": 12,
          "zIndex": 0,
        },
        Object {
          "borderColor": "#E0E0E5",
        },
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "color": "#AA33FF",
          "fontFamily": "Roboto",
        }
      }
      value="foo"
    />
  </View>
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "elevation": 1,
        "position": "absolute",
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 20,
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "overflow": "hidden",
          "width": 40,
        }
      }
    >
      <svg
        data-file-name="SvgClose"
        fill="#FFFFFF"
        height={10}
        testID="Icon"
        width={10}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 8,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#535359",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
    >
      validationMsg
    </Text>
  </View>
</View>
`;
